feat(tokenization): add encode_message to tokenize messages one by one

pco111 opened the pull request:
What does this PR do?
This PR introduces a new method, tokenizer.encode_message, to the base tokenizer class. This method allows for tokenizing a single chat message at a time while correctly handling the conversational context provided by conversation_history. This is particularly useful for token-by-token streaming applications where re-tokenizing the entire conversation history for each new token is inefficient.
The new method works by applying the chat template to the full conversation (history + new message) and then programmatically isolating the tokens that correspond to the new message. This ensures that all special tokens, roles, and formatting are applied correctly according to the model's chat template, maintaining consistency with apply_chat_template.

Fixes #39417
Before submitting
[x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
[x] Did you read the contributor guideline,
Pull Request section?
[x] Was this discussed/approved via a Github issue or the forum? Please add a link
to it if that's the case.
[x] Did you make sure to update the documentation with your changes? Here are the
documentation guidelines, and
here are tips on formatting docstrings.
[x] Did you write any new necessary tests?

Who can review?
@ArthurZucker @Rocketknight1

Commit:
feat(tokenization): add encode_message to tokenize messages one by one

Rocketknight1 commented:
Hi @pco111, this is a cool idea, but I'm not sure about some of the details! In particular, the interaction with `add_generation_prompt` is awkward. If we set that to `True`, then a common scenario is that the `conversation_history` will be tokenized like this:

```
<im_start>user
message<im_end>
<im_start>assistant
```

But in this case, `encode_message()` will treat `<im_start>assistant` as part of the history, and remove it from the encoded message, and then the encoded message will be incomplete. I'm not sure what the best solution is - maybe always set `add_generation_prompt` to `False`?

Commit:
Fix the `encode_message` method, remove the `add_generation_prompt` parameter and add the corresponding error handling. Update the document to reflect this change and verify the error handling in the test.

pco111 commented:
> Hi @pco111, this is a cool idea, but I'm not sure about some of the details! In particular, the interaction with `add_generation_prompt` is awkward. If we set that to `True`, then a common scenario is that the `conversation_history` will be tokenized like this, where the "generation prompt" is the last line:
> 
> ```
> <im_start>user
> message<im_end>
> <im_start>assistant
> ```
> 
> But in this case, `encode_message()` will treat `<im_start>assistant` as part of the history, and remove it from the encoded message, and then the encoded message will be incomplete. I'm not sure what the best solution is - maybe always set `add_generation_prompt` to `False`?



> Hi @Rocketknight1 ,
>
> Thank you so much for your insightful feedback! You've pointed out a very important edge case with add_generation_prompt that I had overlooked.
>
> Following your thoughts, I've opted for a clearer and more robust approach:
>
> 1. Explicitly Disallowed add_generation_prompt: The encode_message method now raises a ValueError if add_generation_prompt is passed. This prevents any ambiguity.
> 2. Updated Documentation: The docstring for encode_message now clearly states that it does not handle the generation prompt and advises users on how to add it separately if needed.
> 3. Updated Tests: The tests have been updated to reflect this new design. There is now a test to ensure that the ValueError is raised correctly.
>
> Thank you again for guiding me toward a better solution! I've pushed the new changes for your review.

Rocketknight1 reviewed (commented):
Made some comments! Also, check the CI on Github - you may need to run `make fixup` to get the style tests to pass.

Commit:
Optimize the `encode_message` method, improve the processing logic of the empty dialogue history, and ensure that the chat template can be applied correctly when the dialogue history is empty. Update the document to reflect these changes.

Commit:
The `_encode_message` method is deleted, the message coding logic is simplified, and the functional integrity of the `encode_message` method is ensured. Update the document to reflect these changes.

Commit:
Docs fix

pco111 commented:
> Made some comments! Also, check the CI on Github - you may need to run `make fixup` to get the style tests to pass.

> Hi @Rocketknight1,
>
> Thank you for the detailed and helpful feedback! I've updated the PR according to all your suggestions:
>
> * The _encode_message helper has been folded into the main encode_message function.
> * An optimization has been added to handle empty conversation history directly.
> * The "Tensor Cores" typo has been corrected.
> * The new tests have been moved into the existing TokenizerUtilsTest class.
>
> All local checks (make fixup and pytest) are passing. The code should be in much better shape now. Thanks again for your guidance!

Commit:
Revert changes in docstring of pad()

Commit:
Revert changes in docstring

pco111 commented:
> Hi @pco111, I think Copilot (or whatever code agent you're using) is making a lot of unrelated changes to the docstring in `get_chat_template`. Can you clean that up and then I'll do a final review?

> Hi @Rocketknight1,
>
> My apologies for the unintended changes in the previous commit. It seems my formatting agent was a bit too aggressive.
>
> I've now cleaned up the docstrings and reverted all unrelated changes. 
>
> Thank you for your patience. Look forward to your final review!

pco111 commented:
> > Hi @pco111, I think Copilot (or whatever code agent you're using) is making a lot of unrelated changes to the docstring in `get_chat_template`. Can you clean that up and then I'll do a final review?
> 
> > Hi @Rocketknight1,
> > My apologies for the unintended changes in the previous commit. It seems my formatting agent was a bit too aggressive.
> > I've now cleaned up the docstrings and reverted all unrelated changes.
> > Thank you for your patience. Look forward to your final review!

Hi @Rocketknight1 , sorry asking again, but just a gentle reminder 😊

Rocketknight1 reviewed (approved):
LGTM now! cc @arthurzucker if you're still happy with it - it's adding a new method to all tokenizers so it probably needs a core maintainer review.

I might also look at refactoring/changing this after [chat schemas](https://github.com/huggingface/transformers/pull/39609) are added, since we might be able to use those to isolate tokens from the final message too.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_39507). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker reviewed (approved):
Fine by me!

pco111 committed:
Update src/transformers/tokenization_utils_base.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

Commit:
Repair the call of the `encode_message` method, update it to `encode_message_with_chat_template` to support the chat template, and adjust the relevant test cases to reflect this change.

Commit:
Optimize the call format of the `apply_chat_template` method, and merge multi-line calls into a single line to improve code readability.

pco111 committed:
Merge branch 'main' into fix/issue-39417-incremental-tokenization

pco111 committed:
Merge branch 'main' into fix/issue-39417-incremental-tokenization
